### ISSUE TITLE
Notes about first steps with MXCuBE web development

### ISSUE THREAD

**Issue by fabcor-maxiv** [mxcube/mxcube3#1047]
Notes taken while trying to build an MXCuBE development environment from scratch, following the [instructions on `mxcube/mxcubeweb` *wiki* page "*Home*"](https://github.com/mxcube/mxcubeweb/wiki).

* Is it mxcube3 or mxcubeweb?

* Version numbers of mxcube3 / mxcubeweb? The following does not make much sense:

  > The underlying beamline control layer is implemented using the library mxcubecore previously known as HardwareRepository. The mxcubecore module is compatible with both MXCuBE3 also known as MXCuBE-Web and the MXCuBE-Qt application. The earlier versions of MXCuBE3 (upto 3.2.x) uses HardwareRepository while versions after 4.x uses mxcubecore.
  >
  > The install instructions for version 3 of MXCUBE-Web can be found here (instructions 3.x)
  >
  > The install instructions for version 4 can be found below.

  https://github.com/mxcube/mxcubeweb/wiki#hardwarerepository-and-mxcubecore

* Document does not say to go into `mxcubeweb` directory for *conda* environment creation command, where the *Yaml* file is located.

* Repo is `mxcubeweb` but import package is `mxcube3`. Why?

  *Note: if MXCuBE is an application then it is quite likely that nothing else needs to import `mxcube3` and it should be possible to safely rename to `mxcubeweb` (or simply `mxcube`?) without breaking others.*

* `poetry install` does not work.

  > $ poetry install
  >
  > Installing dependencies from lock file
  >
  > Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.
  >
  > Because mxcubeweb depends on bcrypt (^4.0.1) which doesn't match any versions, version solving failed.

  *Note: It seems like `poetry lock` helps to get past the issue.
  But is that the correct fix? Is that really what we want?*

* Why do we have a section `D1`? Why suddenly `D`? Is that a typo or a copy-paste from an older document?

* The instruction `cd mxcube3/ui` seems incorrect. First this assumes we are back outside of the repo. And also obviously it should be `cd mxcubeweb/ui`. 

* `npm install` does not work.

  *Note: it seems like `npm install --legacy-peer-deps` could help. But is that really a good solution? Does it break something else? There is still a bunch of warning with "high security vulnerabilities"*

* It is a bit strange to introduce `npm run start` at this point, since it opens a web browser that shows a broken app because indeed we have not started the back end yet.

* Before starting the back end server we should start the Redis server but it is not documented. Also it is not installed at all, it should be in `conda-environment.yaml`.

* Maybe we could say that one can authenticate with `idtest0`. See the *ISPyB* mockup for details about authentication.

**Comment by fabcor-maxiv** [comment]
@marcus-oscarsson @elmjag @meguiraun 

I can likely help with fixing those things. But I rather get advice and feedback from others first.

Additionally I suggest we fix the GitHub workflow to automatically publish the Sphinx doc (should be handled as its own thing), and move those instructions from *wiki* to `/docs/`.

**Comment by marcus-oscarsson** [comment]
Thanks the install instructions are at the moment a bit confusing as we are just transitioning into mxcube-web (from mxcube3) and we have not yet made all the changes everywhere yet. As I mentioned in last our meeting, ill try to cleanup the install instructions and then you can continue to improve what you find is missing.

**Comment by marcus-oscarsson** [comment]
@fabcor-maxiv , Yeah, it would indeed be very nice if we could  build the docs in the CI :)

**Comment by elmjag** [comment]
Regarding auto-building docs, we have discussed this a bit here.

Ideally, we would like that CI builds docs on each push(merge?) to develop branch. Build docs should be published for example to github pages. This should be done for both mxcubeweb and mxcubecore repositories.

We can look into details how to accomplish that, if that's an acceptable plan.

**Comment by marcus-oscarsson** [comment]
Sounds awesome :)

**Comment by marcus-oscarsson** [comment]
So I updated the install instructions to reflect our current situation. I went through the steps on my computer and it worked fine :), but its probably nice if someone else does the same @fabcor-maxiv  :). 

The versioning is indeed a bit odd at the moment, it should perhaps have been 0.x instead of 1.x. The idea was simply that we start with version 4.x once  we had made the majority of the bigger changes we wanted which is more or less the case now. The final step would be to rename the mxcube3 module to mxcubeweb, simply to make the lifes of everybody that needs to rebase on develop. I hope we can do this the coming weeks

**Comment by fabcor-maxiv** [comment]
@marcus-oscarsson I added a couple of edits to the wiki.

**Comment by marcus-oscarsson** [comment]
Thanks !

**Comment by axelboc** [comment]
The updated instructions on the wiki are a lot clearer, thanks @marcus-oscarsson and @fabcor-maxiv!

Couple of notes from my end:

- **MXCuBE-Web** is the name of the whole web project (back-end + front-end), so the name of the Conda environment, `mxcubeweb`, makes sense. However, we should be careful not to refer to the back-end alone as `mxcubeweb`, which makes me think that the folder `mxcube3` should be renamed to `server` instead.
- Eventually, it'd be good if the local installation steps were more in line with the ones used in the CI (and/or the other way around). Currently, the CI uses Mamba, doesn't run poetry, installs LDAP dependencies, etc.
- Could be good to add a section about what do to after pulling new changes to `mxcubeweb` and `mxcubecore` (i.e. update the environment, run `pip install -e .` again, etc.) There's probably something to be improved around mapping commits/versions in `mxcubeweb` to commits/versions in `mxcubecore` (this is important, for instance, when hunting for the source of a regression) -- I've seen git submodules used for this, but it's not the easiest thing to work with... 
- When running the front-end development server, it doesn't seem to make sense to run `mxcubeweb-server` with flag `--static-folder $(pwd)/mxcubeweb/ui/build/`. Could this flag be made optional (assuming it isn't already)?
- Very selfish of me, but long term, it'd be nice to have a minimal set of instructions specifically for front-end developers (skipping things such as `pip install -e .`, `conda-environment-dev.yml`, starting Redis, etc.)
- Finally, would it make sense for the wiki page to be moved into the project? I feel like the introduction could be moved to the `README` and the rest to `CONTRIBUTING`.

**Comment by marcus-oscarsson** [comment]
Thanks @axelboc lots of good ideas. Ill see what we can do

**Comment by fabcor-maxiv** [comment]
@axelboc good points!

* Maybe the document should have the instructions for "dev" installation only. Normal installation (for production for example) is a whole other story, that could deserve its own document. Maybe?
* True that we do not use Poetry. To be honest I find Poetry quite difficult to work with. In particular Poetry is annoying when we want to have a dependency installed as editable -- `mxcubecore` in our case -- or maybe things have changed since last time I seriously looked at Poetry. And since we recommend to use *conda*, which includes the Python virtual environment functionality, the advantage of using Poetry is quite reduced I think. The lockfile feature of Poetry is nice-ish though, maybe that is a good enough reason to keep Poetry?
* `mxcubeweb` repository contains both the web front-end and the web back-end. I would not rename to `server`, `mxcubeweb` seems fine enough for me. Or maybe I misunderstood this point?
* Re-running `pip install --editable` again is typically only necessary when packaging metadata has changed (version string, dependencies, things like that).
* The content of the wiki page could be moved into the README, that is also fine by me. Originally I was planning to move it into the `/docs/`, see https://github.com/mxcube/mxcubeweb/issues/1053. Both are fine. Could even be both (with some trickery).
* To simplify the DX (developer experience) even more, maybe something like [*dev container*](https://containers.dev/) could be an idea to investigate. I do not know. It is a new-ish trendy thing, maybe it falls out of fashion in 6 months, maybe it sticks. I do not know. I am the opinion that we should stick to basic standard tools and then every dev does it the way they like it (personal preferences). We should collect feedback if the first-time dev experience really is too annoying/complex. Not overthink too much, and build scripts and whatnot that are never kept up-to-date.

**Comment by marcus-oscarsson** [comment]
> The updated instructions on the wiki are a lot clearer, thanks @marcus-oscarsson and @fabcor-maxiv!
> 
> Couple of notes from my end:

Thanks :)

> 
>     * **MXCuBE-Web** is the name of the whole web project (back-end + front-end), so the name of the Conda environment, `mxcubeweb`, makes sense. However, we should be careful not to refer to the back-end alone as `mxcubeweb`, which makes me think that the folder `mxcube3` should be renamed to `server` instead.
> 

Sure, I did not really see a problem with naming the python module mxcubeweb but you are right that its important that its clear that the name does not only refer to the backend.

I'm not too much in-favor of the name `server` though its a bit vague in my opinion.  We will try to come up with a better name if you guys prefer something else other than mxcubeweb. Is `mxcubeweb-server` too long, or simply `backend` ? 

>     * Eventually, it'd be good if the local installation steps were more in line with the ones used in the CI (and/or the other way around). Currently, the CI uses Mamba, doesn't run poetry, installs LDAP dependencies, etc.

Actually we could do it exactly the same way, as you mention above I thought it would be easier for people not interested in backend development (just trying the application, or front end developers) to just use poetry install.

I wanted to make it work with simply using `pip install` but if we would like that, we need someway to provide a pre- built front end for that release. I was thinking of adding this in the CI but that somehow felt unconventional ? Like this one could simply `pip install mxubeweb` and it would work.

We should then consider having two or three different instructions one for demo purposes, backend-developers, and one for front end developers ?

> 
>     * Could be good to add a section about what do to after pulling new changes to `mxcubeweb` and `mxcubecore` (i.e. update the environment, run `pip install -e .` again, etc.) There's probably something to be improved around mapping commits/versions in `mxcubeweb` to commits/versions in `mxcubecore` (this is important, for instance, when hunting for the source of a regression) -- I've seen git submodules used for this, but it's not the easiest thing to work with...

Id suggest simply pinning the `mxcubecore` version in the `pyproject.toml`. We have used submodules in the past but it have created lots of difficulties.

> 
>     * When running the front-end development server, it doesn't seem to make sense to run `mxcubeweb-server` with flag `--static-folder $(pwd)/mxcubeweb/ui/build/`. Could this flag be made optional (assuming it isn't already)?
>
It should already be the case :)
 
>     * Very selfish of me, but long term, it'd be nice to have a minimal set of instructions specifically for front-end developers (skipping things such as `pip install -e .`, `conda-environment-dev.yml`, starting Redis, etc.)

I guess the `pip install mxcubeweb` would do the trick then :) Well it wont start redis :)

> 
>     * Finally, would it make sense for the wiki page to be moved into the project? I feel like the introduction could be moved to the `README` and the rest to `CONTRIBUTING`.

Yeah, I think we can manage that.

**Comment by marcus-oscarsson** [comment]
Ahh, we posted at the same time @fabcor-maxiv:

We seem to somehow be converging though:

What I take a way from this:

- We need different install instructions for three possibly even four cases: demo, front end dev, backend dev and production.
- We should consider a different name for the `mxcube3` module other than `mxcubeweb` to avoid name confusion
- We  should make it possible to simply perform `pip install mxcubeweb`
- We should move the install instructions from wiki to docs (and README and CONTRIUBUTING)

Other notes on the above:
We have been switching between different tools now and then from pip to poetry, pyenv, brew, what not, to conda and mamba. I think we are all happy to use new tools if there are good reasons for it i.e they solve problems we already have with the existing ones. But we should also keep in mind that we are busy enough trying to make things run :)

**Comment by marcus-oscarsson** [comment]
@fabcor-maxiv: It would be really cool if you could look at the documentation bit as you mentioned above do you think there is a reasonably easy way to either create or populate README and CONTRIUBUTING automatically ?

@axelboc, @fabcor-maxiv: What do you think of building the front end in the CI, or is there some other cleaver way of doing this.

**Comment by rhfogh** [comment]
A couple of comments:

If going for submodules then please, *please* do not make mxcubecore a submodule of mxcubeweb, or vice versa. It makes it incredibly messy when you are developing both. Rather make a third repository that contains nothing but submodules, and then pull that when you want a fixed version.

It would be nice to have an established setup/update procedure, but I am probably out of step here. I have multiple versions around, both mxcubeqt and mxcubeweb, and want to keep the repositories in fixed locations on my computer, not pull them down new every time before I start working on them, the way you would when getting a release.  Is there, or will there be, a standard procedure that fits with that?

**Comment by marcus-oscarsson** [comment]
We have discussed and used (for a long time) submodules in the past and the decision have been to stop using them, for various reasons. Ill think about it a bit more, I think keeping the `pyproject.toml` file updated with the `mxcubecore` version should be enough for front end development, but its true that its not sufficient for backend development.

@rhfogh Not sure I understood the last paragraph, you mean that you have one (the same) mxcubecore for both qt and web ?

**Comment by axelboc** [comment]
Regarding naming, I was mainly suggesting renaming the folder in the repo currently called `mxcube3` to `server` (or `backend` or whatever, just anything but `mxcubeweb` 😁). Though it is true that naming the module `mxcubeweb` is also confusing. I do like `mxcubeweb-server` since it's consistent with the naming of the script that starts the server.

---

Regarding documentation, I'm of the opinion that maintainer/contributor documentation should not live on a website but in `CONTRIBUTING.md` -- documentation sites should be for users/consumers only. When you're setting up a project, you clone the repo, open your IDE and most likely open a shell in your IDE, so from there, it's easier to open a file than to switch back and forth to the browser.

IMO, the `README` should contain general information about the project (what it's about, who's maintaining it, link to demo/docs, etc.) Some of it can be duplicated on the documentation site, but I don't really see the need to pull one from the other -- it's not like it's supposed to change very often anyway (contrary to contributers documentation, which can evolve more often).

**Comment by marcus-oscarsson** [comment]
Hm, that makes alot of sense. Maybe moving it from the wiki to the `CONTRIBUTING` and `README` is a good start. One can probably include those fairly easy in the "built" documentation if one wants, what do you think about that @fabcor-maxiv

**Comment by fabcor-maxiv** [comment]
> It would be really cool if you could look at the documentation bit as you mentioned above do you think there is a reasonably easy way to either create or populate README and CONTRIUBUTING automatically ?

I would need to look at it closer. If I recall correctly it is possible to setup the Sphinx doc so that it includes the README and CONTRIBUTING documents somehow. I think I did something like this before already. Indeed I agree with @axelboc it would be preferable if the "first steps dev instructions" live in a `CONTRIBUTING` document directly at the root of the repo -- instead of only in `/docs/` and GitHub pages.

**Comment by rhfogh** [comment]
@marcus-oscarsson 
Maybe I did not express myself very well (or set up things very well). I do have separate repository copies of mxcubecore for mxcubeqt and mxcubeweb, just that they live in a fixed location in my home directory. I do not want to download them afresh at any time, but to have my test setups point to that fixed location and change things in those repository copies as needed.

**Comment by marcus-oscarsson** [comment]
@fabcor-maxiv: Ok, I see. Its maybe best to start with whatever is easier, like this we get something in place that works which we can extend. So it also seems like there is an agreement to extend the CONTRIBUTING guidelines with install instructions then.

**Comment by marcus-oscarsson** [comment]
@rhfogh: Ok, sounds reasonable. I think the important thing is that the UI uses the right version of mxcubecore. This is maybe not entirely straight forward as we use `pip install -e` . In this case we will use whatever version that is currently in the mxcubecore direcotry. I think what @axelboc was on to is that it would be nice with some procedure that ensures that there is not a mistake, i.e. an `update` script that checks out the right tag an pefroms `pip install -e.` on mxcubecore  (using the version infomration in mxcubweb) . So, that there is no discrepancy by mistake. 

For front end developers it would suffice with simply pinning the version of `mxcubecore` in `pyproject.toml`. However it becomes more complicated for backend developers.

**Comment by fabcor-maxiv** [comment]
@marcus-oscarsson Not strictly related to this topic, but I guess close enough...

Why are `setuptools`  and `setuptools_scm` list as build dependencies? Is it safe to remove them?

https://github.com/mxcube/mxcubeweb/blob/9cc9e9e3c3f226178c586f40d6a38fd1e07ce4b1/pyproject.toml#L71

Same in `mxcubecore`.

Also do we know that the `conda-install.sh` and `debian-install.sh` scripts are up-to-date? Do we want to keep them?

**Comment by marcus-oscarsson** [comment]
:), they are needed for pip as there is no setup.py. If they are not there we can do pip install -e .

**Comment by fabcor-maxiv** [comment]
I think *pip* and *poetry* both implement the standardized "editable install" (as introduced originally by [PEP 660](https://peps.python.org/pep-0660/)). So there should be no need to do anything special to enable editable installation with `pip install --editable`.

**Comment by marcus-oscarsson** [comment]
You can give it a try, but I dont think you can do a `pip install .` without "setuptools>=45" and "setuptools_scm[toml]>=6.2"

**Comment by marcus-oscarsson** [comment]
It seems however that we can do editable installs with Poetry now which we could not do before,  so if that works well we can remove the usage of pip I believe

**Comment by fabcor-maxiv** [comment]
It definitely should work with pip as well. Unless pip is not up-to-date. I guess *pip >= 21.3* should work.

**Comment by marcus-oscarsson** [comment]
Ok, I trust you on that :), if so I guess its fine to remove them

**Comment by fabcor-maxiv** [comment]
@marcus-oscarsson For info regarding why I asked about `setuptools-scm` (and `setuptools`), this is the issue I had locally that now also happens in the GitHub workflow: https://github.com/fabcor-maxiv/mxcubeweb/actions/runs/6299352605/job/17099938056

_[There is also the issue that Redis is not running, but I guess/hope that it is not related.]_

I will make a pull request to remove `setuptools` and `setuptools-scm` from the dependencies.

**Comment by marcus-oscarsson** [comment]
Okey, go a head :), well you will need redis to run the server. We are already doing it in the test CI pipeline however, do you need to run the server to build the documentation ?

**Comment by fabcor-maxiv** [comment]
Another thing that I noticed (and that is probably not critical for now, in other words it can wait until after code camp to be addressed) is that in our instructions we have something like that:

1. install Python dependencies with conda (`conda-environment.yaml`)
1. install Python dependencies with not conda (via `poetry install` in our case)
1. install Python dependencies with conda (`conda-environment-dev.yaml`)

And as far as I understood, this step 3 is a "*big no-no*". Pip (or Poetry in our case) after conda is fine. But conda after pip (or Poetry) is bad. So far it seems like it has not been an issue for anyone luckily, but I guess we should refrain from encouraging this. Especially I am not too sure what is the point of the `conda-environment-dev.yaml`. When we run `poetry install` we already get things like `pytest` and so on. I do not think we need the step number 3 at all.

A suggestion (hot take, without deep investigation of pros and cons) could be:
* get rid of `conda-environment-dev.yaml`
* step 1 should stay `conda create --file conda-environment.yaml`
* step 2 should be changed to `poetry install --only main`
* step 3 should be changed to `poetry install` (obviously this step is necessary only for development, testing, and other things like that)

**Comment by fabcor-maxiv** [comment]
I think most (if not all) of this has been addressed or fixed one way or another.
